libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -0,0 +1,500 @@
+//! # Buildpack output
+//!
+//! Use the [`BuildpackOutput`] to output structured text as a buildpack is executing.
+//!
+//! ```
+//! use libherokubuildpack::buildpack_output::BuildpackOutput;
+//!
+//! let mut output = BuildpackOutput::new(std::io::stdout())
+//!     .start("Heroku Ruby Buildpack")
+//!     .warning("No Gemfile.lock found");
+//!
+//! output = output
+//!     .section("Ruby version")
+//!     .finish();
+//!
+//! output.finish();
+//! ```
+//!
+use crate::buildpack_output::ansi_escape::{BOLD_PURPLE, CYAN, RED, YELLOW};
+use crate::buildpack_output::util::{prefix_first_rest_lines, prefix_lines, ParagraphInspectWrite};
+use crate::write::line_mapped;
+use std::fmt::Debug;
+use std::io::Write;
+use std::time::Instant;
+
+mod ansi_escape;
+mod duration_format;
+pub mod style;
+mod util;
+
+/// See the module docs for example usage.
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug)]
+pub struct BuildpackOutput<T> {
+    pub(crate) started: Option<Instant>,
+    pub(crate) state: T,
+}
+
+/// Various states for [`BuildpackOutput`] to contain.
+///
+/// The [`BuildpackOutput`] struct acts as an output state machine. These structs
+/// are meant to represent those states.
+#[doc(hidden)]
+pub mod state {
+    use crate::buildpack_output::util::ParagraphInspectWrite;
+    use crate::write::MappedWrite;
+    use std::time::Instant;
+
+    #[derive(Debug)]
+    pub struct NotStarted<W> {
+        pub(crate) write: ParagraphInspectWrite<W>,
+    }
+
+    #[derive(Debug)]
+    pub struct Started<W> {
+        pub(crate) write: ParagraphInspectWrite<W>,
+    }
+
+    #[derive(Debug)]
+    pub struct Section<W> {
+        pub(crate) write: ParagraphInspectWrite<W>,
+    }
+
+    pub struct Stream<W: std::io::Write> {
```

Comment:
Similarly, this still compiles when `pub(crate)` is used:

```suggestion
    pub(crate) struct NotStarted<W> {
        pub(crate) write: ParagraphInspectWrite<W>,
    }

    #[derive(Debug)]
    pub(crate) struct Started<W> {
        pub(crate) write: ParagraphInspectWrite<W>,
    }

    #[derive(Debug)]
    pub(crate) struct Section<W> {
        pub(crate) write: ParagraphInspectWrite<W>,
    }

    pub(crate) struct Stream<W: std::io::Write> {
```

Reply:
Resolved via adding doc tests that exercise these structs